appengine/standard_python37/pubsub/main.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2019 Google, LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START app]
+import base64
+from flask import current_app, Flask, render_template, request
+import json
+import logging
+import os
+
+from google.auth import jwt
+from google.auth.transport import requests
+from google.cloud import pubsub_v1
+from google.oauth2 import id_token
+
+
+app = Flask(__name__)
+
+# Configure the following environment variables via app.yaml
+# This is used in the push request handler to verify that the request came from
+# pubsub and originated from a trusted source.
+app.config['PUBSUB_VERIFICATION_TOKEN'] = \
+    os.environ['PUBSUB_VERIFICATION_TOKEN']
+app.config['PUBSUB_TOPIC'] = os.environ['PUBSUB_TOPIC']
+app.config['GCLOUD_PROJECT'] = os.environ['GOOGLE_CLOUD_PROJECT']
+
+# Global list to store messages, tokens, etc. received by this instance.
+MESSAGES = []
+TOKENS = []
+HEADERS = []
+CLAIMS = []
+
+# [START index]
+@app.route('/', methods=['GET', 'POST'])
+def index():
+    if request.method == 'GET':
+        return render_template('index.html', messages=MESSAGES, tokens=TOKENS,
+                               headers=HEADERS, claims=CLAIMS)
+
+    data = request.form.get('payload', 'Example payload').encode('utf-8')
+
+    publisher = pubsub_v1.PublisherClient()
+    topic_path = publisher.topic_path(app.config['GCLOUD_PROJECT'],
+                                      app.config['PUBSUB_TOPIC'])
+    future = publisher.publish(topic_path, data)
+    future.result()
+    return 'OK', 200
+# [END index]
+
+
+# [START push]
+@app.route('/_ah/push-handlers/receive_messages', methods=['POST'])
+def receive_messages_handler():
+    # Verify that the request originates from the application.
+    if (request.args.get('token', '') !=
+            current_app.config['PUBSUB_VERIFICATION_TOKEN']):
+        return 'Invalid request', 400
+
+    # Verify that the push request originates from Cloud Pub/Sub.
+    try:
+        # Get the Cloud Pub/Sub-generated JWT in the "Authorization" header.
+        bearer_token = request.headers.get('Authorization')
+        token = bearer_token.split(' ')[1]
+        TOKENS.append(token)
+
+        header = jwt.decode_header(token)
```

Comment:
Is this verifying the audience ('aud') field? Does it need to? This makes sure it is from pub/sub, but it could possibly be relayed from a push notification to a different app. Or do all push notifications use the same aud value?

Reply:
The 'aud' field is in the JWT claim. If `--push-auth-token-audience` isn't set when creating the push subscription, then the `aud` field will be set to the push endpoint. `jwt.decode_header` decodes the first section of the JWT; it's a convenient function to showcase the `google.auth.jwt` module but there's not much else to it. If `aud` is set, all push notifications use the same value. 

The checking of the JWT signature happens in `id_token.verify_oauth2_token()`. 

I don't have code that specifically checks the `aud` field, should I add? 

```
# JWT header
{"alg":"RS256","kid":"7d680d8c70d44e947133cbd499ebc1a61c3d5abc","typ":"JWT"}

# JWT claim
{
   "aud":"https://example.com",
   "azp":"113774264463038321964",
   "email":"gae-gcp@appspot.gserviceaccount.com",
   "sub":"113774264463038321964",
   "email_verified":true,
   "exp":1550185935,
   "iat":1550182335,
   "iss":"https://accounts.google.com"
  }
```

Reply:
No, I don't think so. This all looks very good to me. Go ahead and merge when you're ready, or let me know if you want me to do it.

Reply:
@anguillanneuf I'm just getting started reading this, so apologies if you are doing this, but the rule of thumb on these things is to verify the signature before touching any of the fields, then look at the expires and then the audience. (in that order).   Then everything else.

Reply:
@lesv 1. verifying signature and 2. checking expiration time are both done by the Google Auth Python library. I added 3. checking the audience field in the claim. Thanks for catching this!